z3/src/z3_solver.cpp

```diff
@@ -0,0 +1,376 @@
+				"Z3 does not support uninterpreted type with non-zero arity.");
+	}
+
+	return std::make_shared < Z3Sort > (z_sort, ctx);
+}
+
+Sort Z3Solver::make_sort(SortKind sk) const {
+	z3::sort z_sort = ctx.bool_sort();
+
+	if (sk == BOOL) {
+		z_sort = ctx.bool_sort();
+	} else if (sk == INT) {
+		z_sort = ctx.int_sort();
+	} else if (sk == REAL) {
+		z_sort = ctx.real_sort();
+	} else {
+		std::string msg("Can't create sort with sort constructor ");
+		msg += to_string(sk);
+		msg += " and no arguments";
+		throw IncorrectUsageException(msg.c_str());
+	}
+
+	Sort final_sort = std::make_shared < Z3Sort > (z_sort, ctx);
+	return final_sort;
+}
+
+Sort Z3Solver::make_sort(SortKind sk, uint64_t size) const {
+	z3::sort z_sort = ctx.bool_sort();
+	if (sk == BV) {
+		z_sort = ctx.bv_sort(size);
+	} else {
+		std::string msg("Can't create sort with sort constructor ");
+		msg += to_string(sk);
+		msg += " and an integer argument";
+		throw IncorrectUsageException(msg.c_str());
+	}
+	return std::make_shared < Z3Sort > (z_sort, ctx);
+}
+
+Sort Z3Solver::make_sort(SortKind sk, const Sort &sort1) const {
+	throw NotImplementedException(
+			"Smt-switch does not have any sorts that take one sort parameter yet.");
+}
+
+Sort Z3Solver::make_sort(SortKind sk, const Sort &sort1,
+		const Sort &sort2) const {
+	if (sk == ARRAY) {
+		std::shared_ptr<Z3Sort> cidxsort = std::static_pointer_cast < Z3Sort
+				> (sort1);
+		std::shared_ptr<Z3Sort> celemsort = std::static_pointer_cast < Z3Sort
+				> (sort2);
+		return std::make_shared < Z3Sort
+				> (ctx.array_sort(cidxsort->type, celemsort->type), ctx);
+	} else {
+		std::string msg("Can't create sort with sort constructor ");
+		msg += to_string(sk);
+		msg += " and two Sort arguments";
+		throw IncorrectUsageException(msg.c_str());
+	}
+}
+
+Sort Z3Solver::make_sort(SortKind sk, const Sort &sort1, const Sort &sort2,
+		const Sort &sort3) const {
+	throw NotImplementedException(
+			"Smt-switch does not have any sorts that take three sort parameters "
+					"yet.");
+}
+
+Sort Z3Solver::make_sort(SortKind sk, const SortVec &sorts) const {
+	z3::sort final_sort = ctx.bool_sort();
+
+	if (sk == FUNCTION) {
+		if (sorts.size() < 2) {
+			throw IncorrectUsageException(
+					"Function sort must have >=2 sort arguments.");
+		}
+
+		// arity is one less, because last sort is return sort
+		uint32_t arity = sorts.size() - 1;
+
+		std::vector<z3::sort> zsorts;
+		zsorts.reserve(arity);
+		z3::sort z_sort = ctx.bool_sort();
+
+		for (uint32_t i = 0; i < arity; i++) {
+			z_sort = std::static_pointer_cast < Z3Sort > (sorts[i])->type;
+			zsorts.push_back(z_sort);
+		}
+
+		Sort sort = sorts.back();
+		z_sort = std::static_pointer_cast < Z3Sort > (sort)->type;
+
+		const char *c = "throwaway name";
+		z3::symbol func_name = ctx.str_symbol(c);
+		z3::func_decl z_func = ctx.function(func_name, arity, &zsorts[0],
+				z_sort);
+
+		return std::make_shared < Z3Sort > (final_sort, true, z_func, ctx);
```

Comment:
I would move final_sort, rename it so it's less confusing, and put the make_shared and return in here.